glide-core/redis-rs/redis/src/cluster_async/mod.rs

```diff
@@ -1067,6 +1076,62 @@ enum ConnectionCheck<C> {
     RandomConnection,
 }
 
+/// Represents a pipeline command execution context for a specific node
+#[derive(Default)]
+struct NodePipelineContext<C> {
+    /// The pipeline of commands to be executed
+    pipeline: Pipeline,
+    /// The connection to the node
+    connection: C,
+    /// Vector of (command_index, inner_index) pairs tracking command order
+    /// command_index: Position in the original pipeline
+    /// inner_index: Optional sub-index for multi-node operations (e.g. MSET)
+    command_indices: Vec<(usize, Option<usize>)>,
+}
+
+/// Maps node addresses to their pipeline execution contexts
+type NodePipelineMap<C> = HashMap<String, NodePipelineContext<C>>;
+
+impl<C> NodePipelineContext<C> {
+    fn new(connection: C) -> Self {
+        Self {
+            pipeline: Pipeline::new(),
+            connection,
+            command_indices: Vec::new(),
+        }
+    }
+
+    // Adds a command to the pipeline and records its index
+    fn add_command(&mut self, cmd: Cmd, index: usize, inner_index: Option<usize>) {
+        self.pipeline.add_command(cmd);
+        self.command_indices.push((index, inner_index));
+    }
+}
+
+/// Adds a command to the pipeline of the node with the given address
+fn add_command_to_node_pipeline<C>(
+    pipeline_map: &mut NodePipelineMap<C>,
+    address: String,
+    connection: C,
+    cmd: Cmd,
+    index: usize,
+    inner_index: Option<usize>,
+) {
+    pipeline_map
+        .entry(address)
+        .or_insert_with(|| NodePipelineContext::new(connection))
+        .add_command(cmd, index, inner_index);
+}
+
+/// Adds a command to a random existing node in the pipeline
+fn add_to_random_existing_node<C>(pipeline_map: &mut NodePipelineMap<C>, cmd: Cmd, index: usize) {
+    let mut rng = rand::thread_rng();
+    let addresses: Vec<_> = pipeline_map.keys().cloned().collect();
+    let random_address = addresses.choose(&mut rng).unwrap();
+    let context = pipeline_map.get_mut(random_address).unwrap();
```

Comment:
1. Don't use unwrap in production code. If a bug would cause pipeline_map.keys() to be empty it would crash the whole client. Instead, change this function to return a result and return ClientError if no random address is found.
2. Coping all addresses is redundant, you can achieve the same with:
```
    let mut rng = rand::thread_rng();
    if let Some(node_context) = pipeline_map
        .values_mut()
        .choose(&mut rng) {
            node_context.add_command(cmd, index, None);
            Ok(())
    } else {
        // raise error
    }